File: combination-sum/juhui-jeong.ts
// 시간 복잡도: O(2^(target / minCandidate))
// 공간 복잡도: O(target / minCandidate)
function combinationSum(candidates: number[], target: number): number[][] {
  const result: number[][] = [];

  const dfs = (index: number, remain: number, path: number[]) => {
    if (remain === 0) {
      result.push([...path]);
      return;
    }

    if (remain < 0 || index === candidates.length) {
      return;
    }

    const num = candidates[index];

    path.push(num);
    dfs(index, remain - num, path);
    path.pop();

    dfs(index + 1, remain, path);
  };

  dfs(0, target, []);

  return result;
}

File: combination-sum/juhui-jeong.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        // dfs & 백트래킹
        List<List<Integer>> resultList = new ArrayList<>();
        dfs(candidates, target, 0, new ArrayList<>(), resultList);
        return resultList;
    }
    private void dfs(int[] candidates, int target, int start, List<Integer> path, List<List<Integer>> resultList) {
        if (target == 0) {
            resultList.add(new ArrayList<>(path));
            return;
        }

        if (target < 0) {
            return;
        }

        for(int i = start; i < candidates.length; i++) {
            path.add(candidates[i]);
            dfs(candidates, target - candidates[i], i, path, resultList);
            path.remove(path.size() -1);
        }
    }
}
